### PR TITLE
Probe whether the metadata path is canonicalized in Spark (#1725)

### PR DESCRIPTION
**Pull request by xupefei**
## Description

This issue fixes https://github.com/delta-io/delta/issues/1725.

The mechanism of this fix is to call the Spark internal method, which is used to generate metadata columns, to see if it will canonicalize spaces in a crafted path string. If the answer is yes, then we don't need to do anything on the Delta side; otherwise, we manually canonicalize the obtained metadata column.

Why don't use the Spark internal method on `FileToDvDescriptor`, so both sides of the join are either canonicalized or not-canonicalized? Because most Delta methods are expecting a canonicalized path, thus the returned DF must be canonicalized in all cases.

## How was this patch tested?

Existing tests didn't fail.

### CODE REVIEW

**Comment by felipepessoto** [comment]
Is this a duplicate of #1768?

**Comment by xupefei** [comment]
> Is this a duplicate of #1768?

This is a re-submit of #1768. I accidentally closed that one during a branch renaming.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala [hunk 1]
```diff
@@ -386,15 +388,36 @@ object DeletionVectorBitmapGenerator {
...
+   * In Spark 3.4 the file path metadata column is not canonicalized but it is in Spark 3.4.1. To
+   * make Delta Lake works with both Spark versions, we must use Spark's internal path
+   * transformation method to transform our paths here. This method will return a un-canonicalized
+   * path in Spark 3.4 and a canonicalized one in Spark 3.4.1.
+   *
+   * Related issue: https://github.com/delta-io/delta/issues/1725.
+   */
+  private lazy val sparkMetadataFilePathIsCanonicalized: Boolean = {
+    val probeString = "file:/path with space/data.parquet"
+    val row = FileFormat.updateMetadataInternalRow(
```
Comment (vkorukanti) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala:411]: Given this is not a public API, do we want to capture method not found error here and assume it is has the fix to return the path in URI encoded format?

**Review by vkorukanti — APPROVED** [review]
lgtm, one minor comment.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala (hunk 1 above)
Comment (xupefei) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteWithDeletionVectorsHelper.scala]: Good idea. Done!

**Pull request closed without merging**